crwatcher/watcher.go

```diff
@@ -114,20 +116,47 @@ func (cw *CRWatcher) setupHandler(con ResourceController) {
 	cw.handler = cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
 			r := obj.(*unstructured.Unstructured)
+			if cw.Config.Filter != "" && !cw.hasFilter(r) {
+				return
+			}
 			con.ResourceAdded(r)
 		},
 		DeleteFunc: func(obj interface{}) {
 			r := obj.(*unstructured.Unstructured)
+			if cw.Config.Filter != "" && !cw.hasFilter(r) {
+				return
+			}
 			con.ResourceDeleted(r)
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
 			oldR := oldObj.(*unstructured.Unstructured)
 			newR := newObj.(*unstructured.Unstructured)
-			con.ResourceUpdated(oldR, newR)
+			if cw.Config.Filter != "" {
+				cw.filteredUpdate(con, oldR, newR)
+			} else {
+				con.ResourceUpdated(oldR, newR)
+			}
 		},
 	}
 }
 
+// filteredUpdate performs an updated based on the filter status of the given resource. In the case that we have a
+// filter match for the old resource, but not for the new, we should delete the old resource. Either a no-op or delete
+// is valid here, but we've decided on doing a delete.
+func (cw *CRWatcher) filteredUpdate(con ResourceController, oldR *unstructured.Unstructured, newR *unstructured.Unstructured) {
+	if !cw.hasFilter(oldR) && !cw.hasFilter(newR) {
+		return
+	} else if cw.hasFilter(oldR) && !cw.hasFilter(newR) {
+		con.ResourceDeleted(oldR)
+	} else {
+		// This acts as a type of create, so we need to add a new managed release.
+		if !cw.hasFilter(oldR) && cw.hasFilter(newR) {
+			metrics.ManagedReleases.Inc()
```

Comment:
We need to not have metrics in this class. This is a packages that can be imported into other programs and it should not require internal things to happen. In this case we are doing a create so you should just call the ResourceCreated function. The create function already has to handle being called for a resource that already exists.

Reply:
Actually going through this again, this should just call update. We should add logic on the update side so that if it creates a resource during update it will increment the managed releases counter. This will be easy to do on the helm controller, and harder to do on the template controller. But I think that is acceptable.